Make `Queue.enqueue`, `Queue.enqueue_call`,  `Queue.enqueue_at``Queue.parse_args` accept `pipeline` arg, add `Queue.enqueue_many` method

joshcoden opened the pull request:
Resolves #753 

See also: #241

Allow user to pass in a `pipeline` argument to the enqueue methods.

One main use case is the ability for the caller to use their own pipeline and still be able to utilize the dependency logic defined in `enqueue_call`.

-cc @selwin

selwin commented:
@joshcoden may I trouble you to implement `queue.enqueue_many()` that takes advantage of this `pipeline` argument? :)

My thinking is that you'd use this as well so might as well build it into the library for everyone to use.

joshcoden commented:
> @joshcoden may I trouble you to implement `queue.enqueue_many()` that takes advantage of this `pipeline` argument? :)
> 
> My thinking is that you'd use this as well so might as well build it into the library for everyone to use.

I don't mind, but since it wasn't part of my original PR and desire to write tis PR (so thus I don't have an opinionated idea of what the signature and behavior should be), I'm wondering how you would want it implemented.  Some things that come to mind include:

1. Do we need the order of jobs in the input to matter?
2. What does the signature look like?
3.  How about dependency behavior?  Can the jobs in the input depend on each other?  Should I be refactoring it out of `enqueue_call` to be able to be used in multiple places?

selwin commented:
> 1. Do we need the order of jobs in the input to matter?

Since the API would likely involve accepting an iterable of some sort, I think we might as well enqueue jobs based on the order at which they're passed in.

> 2. What does the signature look like?

I'm open to suggestions, but one thing that came to mind is something like this:
```python
JobData = namedtuple('JobData', ['func', 'args', 'kwargs', 'result_ttl', ...])

queue.enqueue_many(
    JobData(f=foo, kwargs={'bar': 'baz'}, result_ttl=60),
    JobData(f=foo, kwargs={'bar': 'baz'}, result_ttl=60),
)
```

`JobData` would support all arguments accepted by [queue.enqueue()](https://github.com/rq/rq/blob/7a1136644732b6b595019d1aea8e54d769664655/rq/queue.py#L313), except for `depends_on`.

This is not final and I'm happy to explore other suggestions.

3. Yes, dealing with job dependencies will be tricky. We may need to create all jobs without enqueueing them first before setting up the dependencies in a single transaction while we monitor for key changes. I think we can leave out job dependency support for now. I'm happy to accept future enhancements to this if someone is willing to work on it.

joshcoden commented:
> > 1. Do we need the order of jobs in the input to matter?
> 
> Since the API would likely involve accepting an iterable of some sort, I think we might as well enqueue jobs based on the order at which they're passed in.
>
Makes sense.
> > 2. What does the signature look like?
> 
> I'm open to suggestions, but one thing that came to mind is something like this:
> 
> ```python
> JobData = namedtuple('JobData', ['func', 'args', 'kwargs', 'result_ttl', ...])
> 
> queue.enqueue_many(
>     JobData(f=foo, kwargs={'bar': 'baz'}, result_ttl=60),
>     JobData(f=foo, kwargs={'bar': 'baz'}, result_ttl=60),
> )
> ```
> 
> `JobData` would support all arguments accepted by [queue.enqueue()](https://github.com/rq/rq/blob/7a1136644732b6b595019d1aea8e54d769664655/rq/queue.py#L313), except for `depends_on`.
> 

Got it.  I actually think you do have to make a separate type if you want the closest thing to "passing multiple `args` and `kwargs`".  

> 
> 3. Yes, dealing with job dependencies will be tricky. We may need to create all jobs without enqueueing them first before setting up the dependencies in a single transaction while we monitor for key changes. I think we can leave out job dependency support for now. I'm happy to accept future enhancements to this if someone is willing to work on it.

Yeah, I don't think it will be particularly hard, but would require some significant refactoring.  I agree though, supporting dependencies for now would be scope creep.  I would probably be willing to work on it, but it's not something I'm actually looking to do pipelined enqueueing at the moment and would rather prioritize something more basic for now.

joshcoden commented:
@selwin Just added my initial implementation and unit tests for `enqueue_many`.  I also refactored the dependency logic and enqueueing out of `enqueue_call` and into its own method, since it would be useful to be able to use that logic elsewhere.  I can revert it if you would rather.  I left this:

https://github.com/rq/rq/blob/6343686d2579bf1d251cb91d08560e2552cef30b/rq/queue.py#L351-L353

Because I didn't think it made sense to pass `depends_on` to the refactored method since it takes a job instance.  But depends_on is not directly stored on the instance, and I wanted to perform the same check as using `depends_on` without having to actual hit the redis connection, and this seemed like the best way available.  Let me know if you have any thoughts on this issue.

selwin commented on rq/queue.py:
```diff
@@ -310,10 +335,66 @@ def create_job(self, func, args=None, kwargs=None, timeout=None,
+        # WatchError is raised in the when the pipeline is executed, that means
+        # something else has modified either the set of dependencies or the
+        # status of one of them. In this case, we simply retry.
+        if job._dependency_id is not None:
```
Can you check with `job._dependency_ids` instead? `job._dependency_id` will be deprecated soon. I'm ok with this way of checking whether this job has dependencies.

selwin commented on rq/queue.py:
```diff
@@ -329,42 +410,46 @@ def enqueue_call(self, func, args=None, kwargs=None, timeout=None,
+    def enqueue_many(
+        self,
+        job_datas,
+        pipeline=None
```
Do we need to accept `pipeline` argument here?

joshcoden replied:
@selwin I did for completeness sake, especially with the original motivation of this PR was adding the ability to pipeline the various enqueue methods, so it was easy enough to implement this.

selwin commented on rq/queue.py:
```diff
@@ -329,42 +410,46 @@ def enqueue_call(self, func, args=None, kwargs=None, timeout=None,
+                if pipeline is None:
+                    pipe.execute()
+                return jobs
+            except WatchError:
```
Do we need to catch `WatchError` here if `depends_on` is always `None`?

joshcoden replied:
I guess not, if `pipeline` is passed in without the `except` it would be raised to them, which is what we are doing here anyway.  I can drop this code.

selwin commented on rq/queue.py:
```diff
@@ -23,6 +24,30 @@ def compact(lst):
     return [item for item in lst if item is not None]
 
 
+class JobEnqueueData(namedtuple('JobEnqueueData', ["func", "args", "kwargs", "timeout",
```
Is there a reason why we're naming this `JobEnqueueData` as opposed to `JobData`? Also open for other ideas.

selwin replied:
Thinking further about this, you're right that `JobEnqueueData` makes more sense since this is not just simply data about job itself. Just wondering if there's a better name for this. Some ideas:
```python
queue.enqueue_many(
    queue.EnqueueData(func=foo, result_ttl=60),
    queue.EnqueueData(func=bar)
)
```

queue.enqueue_many(
    queue.prepare_data(func=foo, result_ttl=60),
    queue.prepare_data(func=bar)
)
```

joshcoden replied:
I'm fine with either.  You could argue the latter one is better because it does not expose the inner type.

selwin replied:
Ok, then let's go with 
```python
queue.enqueue_many(
    queue.prepare_data(func=foo, result_ttl=60),
    queue.prepare_data(func=bar)
)
```

Also open to suggestion on a better name for `prepare_data`.

joshcoden replied:
@selwin I don't have any strong feelings on the name, `prepare_data` honestly does not seem bad to me.  I'll implement it that way, at least for now.

selwin commented on tests/test_queue.py:
```diff
@@ -507,6 +510,114 @@ def test_enqueue_job_with_dependency(self):
+                    self.assertEqual(len(q.deferred_job_registry), 0)
+                    pipe.execute()
+                    break
+                except WatchError:
+                    continue
```
We don't need this in the test case.

selwin commented on tests/test_queue.py:
```diff
@@ -507,6 +510,114 @@ def test_enqueue_job_with_dependency(self):
+                    self.assertEqual(job.get_status(refresh=False), JobStatus.QUEUED)
+                    pipe.execute()
+                    break
+                except WatchError:
+                    continue
```
We don't need this in the test case I think.

selwin commented:
Also, we'll need to document this feature so it's discoverable. The docs are in the `docs` directory.

joshcoden commented:
> Also, we'll need to document this feature so it's discoverable. The docs are in the `docs` directory.

@selwin I addressed your comments and added some docs, let me know if it's in the right place and if there is anything missing or needing changes.  I assumed it would be in version 1.9.0 when specifying when this new feature was available in the docs.

selwin commented on rq/queue.py:
```diff
@@ -310,10 +322,64 @@ def create_job(self, func, args=None, kwargs=None, timeout=None,
 
         return job
 
+    def setup_dependencies_and_enqueue(
```
One more small change, can you change this to `setup_dependencies` and leave out the `enqueue_job` line? It's cleaner to have a single method to setup dependencies and we can potentially reuse this in other places.

selwin replied:
Unless the `enqueue_job` is part of the logic and it's hard to untangle, in which case we can leave it as it is.

joshcoden replied:
@selwin I thought about doing that, but that means we could have the case if `depends_on` is True and a job is found we would have that job that may already be in the registry of the queue, or the job got no where at all

joshcoden replied:
@selwin Actually, on second thought, there should be a way to refactor it out.  Maybe we should have a discussion on the best way how?   Because if aren't enqueueing we would need to know if the job was deferred or if none were found and should just be enqueued maybe via a bool return value?

selwin replied:
What if we check the job status? If `job.status` is not `DEFERRED`, then it should be enqueued.

joshcoden replied:
@selwin Done.  Should I be increasing the version number to `1.9.0`?:
https://github.com/rq/rq/blob/5620f2932cffb31df58a164d78e5cd4327b89414/rq/version.py#L5

selwin replied:
Yes, would be great if you could do that. After this is merged in, I'll also pull in https://github.com/rq/rq/pull/1480 and cut a release in a few weeks.

joshcoden replied:
@selwin Awesome, thank you so much!  I have updated the version to 1.9.0.

selwin commented:
Thanks!
